### PR TITLE
PlaceGuide List Representation

### PR DESCRIPTION
**Pull request by denniswillie**
In this pull request, I added 3 new script files, which are PlaceGuideOnList.js, ListPlaceGuideDisplayer.js, and UserRepresentation.js. I also add resize listener for when the window is resized so that the map will also be resized.

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/BookmarkPlaceGuideServlet.java [hunk 1]
```diff
@@ -20,17 +20,18 @@
 
   private static final String PLACE_GUIDE_ID_PARAMETER = "placeGuideId";
   private static final String BOOKMARK_HANDLING_TYPE_PARAMETER = "bookmarkHandlingType";
+  private static final String CURRENT_URL_PARAMETER = "currentUrl";
```
Comment (bori00) [inline src/main/java/com/google/sps/servlets/BookmarkPlaceGuideServlet.java]: Where do you use this variable?

File: src/main/webapp/createPlaceGuide.html [hunk 2]
```diff
@@ -46,7 +48,7 @@
             <input type = "checkbox" name = "deleteImage" value = "delete" id = "deleteImage">
             <br>
             <label for = "description">Description:</label><br>
-            <textarea name = "description" rows = "4" cols = "50" id = "description" maxlength = "200" pattern = "[A-Za-z0-9,.\s]+" title = "You can only enter letters, numbers, periods, and comas!"></textarea>
+            <textarea name = "description" rows = "4" cols = "50" id = "description" maxlength = "1000" pattern = "[A-Za-z0-9,.\s]+" title = "You can only enter letters, numbers, periods, and comas!"></textarea>
```
Comment (bori00) [inline src/main/webapp/createPlaceGuide.html]: Please don't set it to 1000 as long as we can't display such a long text.

File: src/main/webapp/placeGuides/PlaceGuideManager.js [hunk 3]
```diff
@@ -6,18 +6,18 @@
  * classes which are responsible for handling the events.
  */
 class PlaceGuideManager {
-  constructor(placeGuideRepository) {
+  constructor(placeGuideRepository, placeGuideDisplayType, hasSubtitle) {
```
Comment (bori00) [inline src/main/webapp/placeGuides/PlaceGuideManager.js]: Let's not add fields to the constructor that we will need to remove anyway. They cause unnecessary conflicts and dependencies.
You can use some dummy values for the title and the subtitle, or you can merge my branch into yours to get the updated version of PlaceGuideManager.

File: src/main/webapp/placeGuides/ListPlaceGuideDisplayer.js [hunk 4]
```diff
@@ -2,75 +2,132 @@
...
+      listSubTitle: "in selected map area"
+    },
+    "CREATED_ALL_IN_MAP_AREA": {
+      listTitle: "My Guides",
+      listSubTitle: "in selected map area"
+    },
+    "BOOKMARKED": {
+      listTitle: "Bookmarked Guides",
+      listSubTitle: ""    }
+  };
```
Comment (bori00) [inline src/main/webapp/placeGuides/ListPlaceGuideDisplayer.js]: I think the ListPlaceGuideDisplayer should not work based on query types, but rather based on the page. We may have different subtitle on pages on which the same query is applied and vice-versa.
Please modify this enum.

File: src/main/webapp/placeGuides/ListPlaceGuideDisplayer.js [hunk 5]
```diff
@@ -2,75 +2,132 @@
...
+  get listPlaceGuideDisplayerDiv() {
+    return this._listPlaceGuideDisplayerDiv;
+  }
 
+  update(placeGuides) {
+    this.removeAllPlaceGuidesFromList();
+    this._listPlaceGuideDisplayerDiv.appendChild(
+        this.createListTitle(
+            ListPlaceGuideDisplayer.QUERY[this._placeGuideDisplayQuery]));
+    this.addPlaceGuidesToList(placeGuides);
```
Comment (bori00) [inline src/main/webapp/placeGuides/ListPlaceGuideDisplayer.js]: Consider removing only the ones which don't need to be added again. And then add only the ones which were not there previously.
For a larger number of placeguides I think it may significantly increase the efficiency. Plus, it's a little annoying that whenever the map bounds are changed, the order of placeguides on the list is completely changed. Instead, We could just append the new ones at the end.

File: src/main/webapp/placeGuides/ListPlaceGuideDisplayer.js [hunk 6]
```diff
@@ -2,75 +2,132 @@
...
-  addToList(placeGuideId) {
+  get listPlaceGuideDisplayerDiv() {
+    return this._listPlaceGuideDisplayerDiv;
+  }
 
+  update(placeGuides) {
+    this.removeAllPlaceGuidesFromList();
+    this._listPlaceGuideDisplayerDiv.appendChild(
+        this.createListTitle(
+            ListPlaceGuideDisplayer.QUERY[this._placeGuideDisplayQuery]));
```
Comment (bori00) [inline src/main/webapp/placeGuides/ListPlaceGuideDisplayer.js]: Don't remove the header each time we update the placeguides.

File: src/main/webapp/placeGuides/ListPlaceGuideDisplayer.js [hunk 7]
```diff
@@ -2,75 +2,132 @@
...
       }
     }
   }
 
   remove(placeGuideId) {
-    // Delete from list.
-    this.removeFromList(placeGuideIde);
-    // Delete from memory.
-    delete this._placeGuidesOnList[placeGuideId];
+    const placeGuideDivId = "placeGuideOnList-" + "{" + placeGuideId + "}";
```
Comment (bori00) [inline src/main/webapp/placeGuides/ListPlaceGuideDisplayer.js]: I would suggest extracting the part for generating the div id in a separate function, assuming that we may need it multiple times. It's less error-prone.

File: src/main/webapp/placeGuides/ListPlaceGuideDisplayer.js [hunk 8]
```diff
@@ -2,75 +2,132 @@
...
+      this._listPlaceGuideDisplayerDiv.removeChild(placeGuideDiv);
+    }
+    return placeGuideDiv;
   }
 
+   // Move the highlighted place guide to top of list.
   highlight(placeGuideId) {
-
+    const placeGuideDiv = this.remove(placeGuideId);
+    this._listPlaceGuideDisplayerDiv.removeChild(document.getElementById("listTitle"));
```
Comment (bori00) [inline src/main/webapp/placeGuides/ListPlaceGuideDisplayer.js]: Again, don't remove the title. Find the next sibling and insert before that one for example. Or create a separate parent div for the placeguides which does not contain the title.

File: src/main/webapp/user/UserRepresentation.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    if (user.publicPortfolio) {
+      var url = new URL("/portfolio.html", document.URL);
+      url.searchParams.append('user', user)
+      window.location = url;
+    } else {
+      alert("The creator of this place guide does not have a private place guide!");
+    }
+  });
+  return userDiv;
+}
```
Comment (bori00) [inline src/main/webapp/user/UserRepresentation.js]: This won't work. First, /portfolio.html will show the user's own portfolio. Plus, UserRepresentation doesn't display the user icon or anything related to the user. Let's leave it for a separate PR.
You can keep a sketch in order not to break other code.

File: src/test/java/com/google/sps/PlaceGuideServletTest.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+  @Before
+  public void setUp() {
+    helper.setUp();
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+}
```
Comment (bori00) [inline src/test/java/com/google/sps/PlaceGuideServletTest.java]: Leave these tests for another PR. They are not finished and they are also not connected to the other parts of the PR.

File: src/main/java/com/google/sps/servlets/BookmarkPlaceGuideServlet.java (hunk 1 above)
Comment (denniswillie) [reply src/main/java/com/google/sps/servlets/BookmarkPlaceGuideServlet.java]: deleted the variable.

File: src/main/webapp/createPlaceGuide.html (hunk 2 above)
Comment (denniswillie) [reply src/main/webapp/createPlaceGuide.html]: set the maxlength to 100.

File: src/main/webapp/placeGuides/PlaceGuideManager.js (hunk 3 above)
Comment (denniswillie) [reply src/main/webapp/placeGuides/PlaceGuideManager.js]: merged your branch with mine.

File: src/main/webapp/placeGuides/ListPlaceGuideDisplayer.js (hunk 4 above)
Comment (denniswillie) [reply src/main/webapp/placeGuides/ListPlaceGuideDisplayer.js]: modified enum

File: src/main/webapp/placeGuides/ListPlaceGuideDisplayer.js (hunk 5 above)
Comment (denniswillie) [reply src/main/webapp/placeGuides/ListPlaceGuideDisplayer.js]: modified the update method.

File: src/main/webapp/placeGuides/ListPlaceGuideDisplayer.js (hunk 6 above)
Comment (denniswillie) [reply src/main/webapp/placeGuides/ListPlaceGuideDisplayer.js]: modified the update method to note remove the header on updating the place guides.

File: src/main/webapp/placeGuides/ListPlaceGuideDisplayer.js (hunk 7 above)
Comment (denniswillie) [reply src/main/webapp/placeGuides/ListPlaceGuideDisplayer.js]: extracted a new function to accommodate getting the placeGuideOnListDivId.

File: src/main/webapp/placeGuides/ListPlaceGuideDisplayer.js (hunk 8 above)
Comment (denniswillie) [reply src/main/webapp/placeGuides/ListPlaceGuideDisplayer.js]: Modified the function so that it doesn't remove the title when highlighting, instead, it will just append the highlighted div after the title.

File: src/main/webapp/user/UserRepresentation.js (hunk 9 above)
Comment (denniswillie) [reply src/main/webapp/user/UserRepresentation.js]: removed user representation.

File: src/test/java/com/google/sps/PlaceGuideServletTest.java (hunk 10 above)
Comment (denniswillie) [reply src/test/java/com/google/sps/PlaceGuideServletTest.java]: removed.

File: src/main/webapp/forms/placeGuideForm.js [hunk 11]
```diff
@@ -70,7 +71,7 @@ async function getFetchedList() {
 }
```
Comment (bori00) [inline src/main/webapp/forms/placeGuideForm.js:33]: Do we still need this?

File: src/main/webapp/forms/placeGuideForm.js [hunk 12]
```diff
@@ -137,3 +138,36 @@ function fillFormWithPlaceGuideData(placeGuide) {
...
+    }
+    document.getElementById('length').value = GET['length'];
+    document.getElementById('isPublic').value = GET['isPublic'];
+    if (GET['placeName'] != 'null') {
+      document.getElementById('placeName').value = GET['placeName'];
+    }
+    document.getElementById('latitude').value = GET['latitude'];
+    document.getElementById('longitude').value = GET['longitude'];
+  }
+}
```
Comment (bori00) [inline src/main/webapp/forms/placeGuideForm.js:149]: Don't you need to use the MDC objects(as initialised above) for setting the values? Are the text input labels, textarea labels, etc. working correctly if you modify directly the value of the inputs instead of setting the value of the MDC objects?

File: src/main/webapp/pages/bookmarkedPlaceGuides_initPage.js [hunk 13]
```diff
@@ -8,6 +8,7 @@
  */
 let map;
 let placeGuideManager;
+const PLACE_GUIDE_DISPLAY_TYPE = 'Bookmarked Guides';
```
Comment (bori00) [inline src/main/webapp/pages/bookmarkedPlaceGuides_initPage.js]: Remove this variable.

File: src/main/webapp/pages/discover_initPage.js [hunk 14]
```diff
@@ -7,6 +7,7 @@
  */
 let map;
 let placeGuideManager;
+const PLACE_GUIDE_DISPLAY_TYPE = 'Discover Guides';
```
Comment (bori00) [inline src/main/webapp/pages/discover_initPage.js]: This too.

File: src/main/webapp/pages/myPlaceGuides_initPage.js [hunk 15]
```diff
@@ -8,6 +8,7 @@
  */
 let map;
 let placeGuideManager;
+const PLACE_GUIDE_DISPLAY_TYPE = 'My Guides';
```
Comment (bori00) [inline src/main/webapp/pages/myPlaceGuides_initPage.js]: And this.

File: src/main/webapp/placeGuides/ListPlaceGuideDisplayer.js [hunk 16]
```diff
@@ -2,75 +2,163 @@
...
+        var placeGuideDivToBeRemoved = placeGuideDiv;
+        placeGuideDiv = placeGuideDiv.nextSibling;
+        this._listPlaceGuideDisplayerDiv.removeChild(document.getElementById(placeGuideDivToBeRemoved.id));
+      } else {
+        delete placeGuidesCopy[this.extractIdFromDivId(placeGuideDiv.id)];
+        placeGuideDiv = placeGuideDiv.nextSibling;
       }
     }
+    this.addPlaceGuidesToList(placeGuidesCopy);
   }
```
Comment (bori00) [inline src/main/webapp/placeGuides/ListPlaceGuideDisplayer.js:61]: Much better! Thanks!

File: src/main/webapp/placeGuides/PlaceGuideManager.js [hunk 17]
```diff
@@ -10,18 +10,22 @@ class PlaceGuideManager {
     DISCOVER: {
       query: PlaceGuideRepository.QUERY_TYPE.ALL_PUBLIC_IN_MAP_AREA,
       guideBookmarkStatusChanged: undefined,
+      name: "DISCOVER"
```
Comment (bori00) [inline src/main/webapp/placeGuides/PlaceGuideManager.js:13]: Where do we need this name?

File: src/test/java/com/google/sps/BookmarkPlaceGuideServletTest.java [hunk 18]
```diff
@@ -0,0 +1 @@
+
```
Comment (bori00) [inline src/test/java/com/google/sps/BookmarkPlaceGuideServletTest.java]: Why do you want to merge to master an empty file?

File: src/test/java/com/google/sps/DeletePlaceGuideServletTest.java [hunk 19]
```diff
@@ -0,0 +1 @@
+
```
Comment (bori00) [inline src/test/java/com/google/sps/DeletePlaceGuideServletTest.java]: Same.

Comment (denniswillie) [reply src/test/java/com/google/sps/DeletePlaceGuideServletTest.java]: removed the file.

File: src/test/java/com/google/sps/BookmarkPlaceGuideServletTest.java (hunk 18 above)
Comment (denniswillie) [reply src/test/java/com/google/sps/BookmarkPlaceGuideServletTest.java]: removed the file.

File: src/main/webapp/pages/bookmarkedPlaceGuides_initPage.js (hunk 13 above)
Comment (denniswillie) [reply src/main/webapp/pages/bookmarkedPlaceGuides_initPage.js]: removed.

File: src/main/webapp/pages/discover_initPage.js (hunk 14 above)
Comment (denniswillie) [reply src/main/webapp/pages/discover_initPage.js]: removed.

File: src/main/webapp/pages/myPlaceGuides_initPage.js (hunk 15 above)
Comment (denniswillie) [reply src/main/webapp/pages/myPlaceGuides_initPage.js]: removed.

File: src/main/webapp/forms/placeGuideForm.js (hunk 11 above)
Comment (denniswillie) [reply src/main/webapp/forms/placeGuideForm.js:33]: removed all functions used for testing.

File: src/main/webapp/placeGuides/PlaceGuideManager.js (hunk 17 above)
Comment (denniswillie) [reply src/main/webapp/placeGuides/PlaceGuideManager.js:13]: for the ListPlaceGuideDisplayer.js

File: src/main/webapp/forms/placeGuideForm.js (hunk 12 above)
Comment (denniswillie) [reply src/main/webapp/forms/placeGuideForm.js:149]: Used the MDC objects for setting the values.

File: src/main/webapp/placeGuides/PlaceGuideManager.js (hunk 17 above)
Comment (bori00) [reply src/main/webapp/placeGuides/PlaceGuideManager.js:13]: Ok, I see.

File: src/main/webapp/createPlaceGuide.html [hunk 20]
```diff
@@ -28,7 +28,7 @@ <h2 class="primary-text align-center">New Audio Guide</h2>
             <input type="hidden" id="placeId" name="placeId">
 
             <label id= "nameInput" class="mdc-text-field mdc-text-field--outlined form-field">
-                <input type="text" class="mdc-text-field__input" aria-labelledby="my-label-id" name = "name" required id = "name" pattern = "[A-Za-z':-\s]+" title = "Enter appropriate name!" maxlength = "20">
+                <input type="text" class="mdc-text-field__input" aria-labelledby="my-label-id" name = "name" required id = "name" pattern = "[A-Za-z\s]+" title = "Enter appropriate name!" maxlength = "20">
```
Comment (bori00) [inline src/main/webapp/createPlaceGuide.html:31]: Why did you remove those characters? The name is actually a title of the guide, restricting it to letters only seems too strict.

File: src/main/webapp/forms/placeGuideForm.js [hunk 21]
```diff
@@ -100,42 +59,39 @@ function fillFormWithPlaceGuideData(placeGuide) {
...
-    document.getElementById('no-img-icon').style.display = 'none';
-    document.getElementById('clear-img-icon').style.display = 'block';
-    activateRemoveImageFeature(true);
-  } else {
-    activateRemoveImageFeature(false);
+    setBlobKeySrcToElement(
+        placeGuide.imageKey,
+        'imagePreview', true);
+    document.getElementById('no-img-icon')
+        .style.display = 'none';
```
Comment (bori00) [inline src/main/webapp/forms/placeGuideForm.js:93]: What about the remove image feature?

File: src/main/webapp/forms/placeGuideForm.js [hunk 22]
```diff
@@ -152,22 +110,40 @@ function fillFormWithPlaceGuideToEdit() {
...
+        new mdc.textField.MDCTextField(document.getElementById('nameInput')),
+        GET['name']);
+    document.getElementById('audioPlayer').src = PlaceGuideOnList.getBlobSrc(GET['audioKey']);
     if (GET['imageKey'] != 'undefined') {
-      document.getElementById('imagePreview').src = getBlobSrc(GET['imageKey']);
+      document.getElementById('imagePreview').style.display = 'block';
+      document.getElementById('imagePreview').src =
+        PlaceGuideOnList.getBlobSrc(GET['imageKey']);
+      document.getElementById('no-img-icon')
+          .style.display = 'none';
```
Comment (bori00) [inline src/main/webapp/forms/placeGuideForm.js:125]: Remove image feature? The clear-img-icon should show up

File: src/main/webapp/forms/setUpFormBlobstore.js [hunk 23]
```diff
@@ -36,6 +36,40 @@ function setFormActionUrl(uploadUrl, formId) {
...
+        preview.setAttribute('src', this.result);
+      });
+
+      reader.readAsDataURL(file);
+    } else {
+      preview.style.display = 'none';
+    }
+  });
+}
+
```
Comment (bori00) [inline src/main/webapp/forms/setUpFormBlobstore.js:72]: These functions were in formUtilities.js, right? Why did you move them back?

File: src/main/webapp/pages/createPlaceGuide_initPage.js [hunk 24]
```diff
@@ -25,8 +25,13 @@ function initPage() {
...
       map = mapWidget.map;
-      placeGuideManager = new PlaceGuideManager(
-          PlaceGuideManager.PAGE.CREATE_PLACE_GUIDE, map);
+      const placeGuideRepository =
+          new PlaceGuideRepository(
+              PlaceGuideRepository.QUERY_TYPE.CREATED_ALL_IN_MAP_AREA);
+      placeGuideManager = new PlaceGuideManager(PlaceGuideManager.PAGE.CREATE_PLACE_GUIDE, map);
+      google.maps.event.addListener(map, 'idle', function() {
+        placeGuideManager.refreshPlaceGuides(map.getBounds(), map.getZoom(), false);
+      });
```
Comment (bori00) [inline src/main/webapp/pages/createPlaceGuide_initPage.js:35]: The listener to the map's idle state was move dto PlaceGuideManager. And it also handles initialising the PlaceGuideRepository.
Why did you move these back to initPage()?
